modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -1180,6 +1199,88 @@
+  including in a `$setWindowFields` output block."
+  [[output-name input-name]]
+  {output-name
+   {$sum input-name
+    "window" {"documents" ["unbounded" "current"]}}})
+
+(defn- sort-lookup
+  "Generates a lookup string for a particular field"
+  [id name]
+  (if (id name)
+    (str "_id." name)
+    name))
+
+(defn- make-window-sort
+  "Converts a `$sort` body to something that can be used in a `sortBy` clause in a
+  `$setWindowFields` stage."
+  [id pairs]
+  (when-let [pair-seq (seq pairs)]
+    (into (ordered-map/ordered-map)
+          (map (fn [[name dir]] [(sort-lookup id name) dir]))
+          pair-seq)))
+
+(defn- get-window-sort-and-partitions
+  "Calculates the appropriate sort and partition fields for a `$setWindowFields` stage."
+  [id breakouts order-by]
+  (let [finest-temporal-index
+        (qp.util.transformations.nest-breakouts/finest-temporal-breakout-index breakouts 2)
+
+        sort-index (or finest-temporal-index
+                       (dec (count breakouts)))
+        sort-name (first (nth (seq id) sort-index))
+        default-sort {(sort-lookup id sort-name) 1}
+        user-sort (when order-by
+                    (binding [*field-mappings*
+                              (merge *field-mappings*
+                                     (into {} (map (juxt identity field-alias)) breakouts))]
+                      (order-by->$sort order-by)))
+        sort-expr (cond
+                    ;; if there is only one breakout, always use the user's sort order
+                    (= (count id) 1)
+                    (or (make-window-sort id user-sort) default-sort)
+
+                    ;; if we don't have a temporal breakout, sort by the last breakout, but
+                    ;; use the user's sort direction if specified
+                    (not finest-temporal-index)
+                    (or (->> user-sort
+                             (filter #(= sort-name (first %)))
+                             (make-window-sort id))
+                        default-sort)
+
+                    :else default-sort)
+
+        partition-expr (into (ordered-map/ordered-map)
+                             (map (fn [[name]] [name (str "$_id." name)]))
+                             (m/remove-nth sort-index id))]
+    {:sort-expr sort-expr
+     :partition-expr partition-expr}))
+
+(defn- generate-window-accumulators
+  "Takes a map of {output-name input-name ...} and generates a `$setWindowFields` stage that
+  produces a cumulative sum of those fields."
+  [window-vals id breakouts order-by]
+  ;; if id is empty, we don't have any breakouts and so don't need to fiddle around with $setWindowFields
+  (if (empty? id)
+    [{$addFields window-vals}]
+    (let [{:keys [sort-expr partition-expr]} (get-window-sort-and-partitions id breakouts order-by)]
+      [{$setWindowFields
+        (cond-> {"sortBy" sort-expr
+                 "output" (into {} (map generate-window-output-clause) window-vals)}
```

Comment:
I'm struggling to read this properly in my head, but it looks like this is:
- `window-vals` maps `output-name` to `input-name`
- `generate-window-output-clause` gets the **pair** `[output-name input-name]`
- ... and returns `{output-name {blah input-name}}` or similar
- which get squashed into a `{}` by `into`
which I think is correct (and probably so if tests are passing).

But it's a lot slicker to write it as
```
(defn window-output-clause
  [input-name]
  {$sum input-name
   "window" {"documents" ["unbounded" "current"]}})
```
and
```suggestion
                 "output" (update-vals window-vals window-output-clause)}
```

Reply:
definitely cleaner